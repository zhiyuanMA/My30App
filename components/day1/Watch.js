import React, { Component } from 'react';

export default class Watch extends Component {}
