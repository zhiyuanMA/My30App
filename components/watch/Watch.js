import React, { Component } from 'react';
import { Platform, View, StatusBar, StyleSheet } from 'react-native';

import WatchRecord from './WatchRecord';
import WatchFace from './WatchFace';
import WatchControl from './WatchControl';

let interval;
const initialState = {
    stopWatch: false,
    resetWatch: true,
    initialTime: 0,
    currentTime: 0,
    recordTime: 0,
    timeAccumulation: 0,
    totalTime: '00:00.00',
    sectionTime: '00:00.00',
    recordCounter: 0,
    records: [],
};
export default class Watch extends Component {
    constructor() {
        super();
        this.state = { ...initialState };
    }

    componentWillUnmount() {
        this._stopWatch();
        this._clearRecord();
    }

    componentDidMount() {
        if (Platform.OS === 'ios') {
            StatusBar.setBarStyle(0);
        }
    }

    _startWatch() {
        if (this.state.resetWatch) {
            this.setState({
                stopWatch: false,
                resetWatch: false,
                timeAccumulation: 0,
                initialTime: new Date().getTime(),
            });
        } else {
            this.setState({
                stopWatch: false,
                initialTime: new Date().getTime(),
            });
        }

        let milSecond, second, minute, countingTime, secmilSecond, secsecond, secminute, seccountingTime;
        interval = setInterval(() => {
            this.setState({
                currentTime: new Date().getTime(),
            });
            countingTime = this.state.timeAccumulation + this.state.currentTime - this.state.initialTime;
            minute = Math.floor(countingTime / (60 * 1000));
            second = Math.floor((countingTime - 6000 * minute) / 1000);
            milSecond = Math.floor((countingTime % 1000) / 10);
            seccountingTime = countingTime - this.state.recordTime;
            secminute = Math.floor(seccountingTime / (60 * 1000));
            secsecond = Math.floor((seccountingTime - 6000 * secminute) / 1000);
            secmilSecond = Math.floor((seccountingTime % 1000) / 10);
            this.setState({
                totalTime:
                    (minute < 10 ? '0' + minute : minute) +
                    ':' +
                    (second < 10 ? '0' + second : second) +
                    '.' +
                    (milSecond < 10 ? '0' + milSecond : milSecond),
                sectionTime:
                    (secminute < 10 ? '0' + secminute : secminute) +
                    ':' +
                    (secsecond < 10 ? '0' + secsecond : secsecond) +
                    '.' +
                    (secmilSecond < 10 ? '0' + secmilSecond : secmilSecond),
            });
            if (this.state.stopWatch) {
                clearInterval(interval);
                this.setState({
                    timeAccumulation: countingTime,
                });
            }
        }, 10);
    }

    _stopWatch() {
        this.setState({
            stopWatch: true,
        });
        clearInterval(interval);
    }

    _addRecord() {
        let { recordCounter } = this.state;
        let counter = recordCounter;
        counter++;
        const colors = ['#FF0000', '#FF00FF', '#00FF00', '#808080', '#FFFF00', '#0000FF'];
        const index = counter % colors.length;
        this.setState({
            recordTime: this.state.timeAccumulation + this.state.currentTime - this.state.initialTime,
            recordCounter: this.state.recordCounter + 1,
            records: [
                { key: counter, title: 'Lab' + counter, time: this.state.sectionTime, color: colors[index] },
                ...this.state.records,
            ],
        });
    }

    _clearRecord() {
        this.state = { ...initialState };
    }

    render() {
        return (
            <View style={styles.watchContainer}>
                <WatchFace totalTime={this.state.totalTime} sectionTime={this.state.sectionTime} />
                <WatchControl
                    addRecord={() => this._addRecord()}
                    clearRecord={() => this._clearRecord()}
                    startWatch={() => this._startWatch()}
                    stopWatch={() => this._stopWatch()}
                />
                <WatchRecord records={this.state.records} />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    watchContainer: {
        alignItems: 'center',
        backgroundColor: '#f3f3f3',
    },
});
