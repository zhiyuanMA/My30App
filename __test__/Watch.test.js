import React from 'react';
import Watch from '../components/watch/Watch';

import renderer from 'react-test-renderer';

it('renders Watch without crashing', () => {
    const rendered = renderer.create(<Watch />).toJSON();
    expect(rendered).toMatchSnapshot();
});
