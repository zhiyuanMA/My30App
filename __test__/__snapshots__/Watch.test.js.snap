// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Watch without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f3f3f3",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderBottomColor": "#ddd",
        "borderBottomWidth": 1,
        "height": 170,
        "paddingBottom": 20,
        "paddingLeft": 30,
        "paddingRight": 30,
        "paddingTop": 20,
        "width": 750,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#555",
            "fontSize": 20,
            "fontWeight": "100",
          }
        }
      >
        00:00.00
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#222",
            "fontSize": 70,
            "fontWeight": "100",
          }
        }
      >
        00:00.00
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#f3f3f3",
        "flexDirection": "row",
        "height": 100,
        "paddingBottom": 0,
        "paddingLeft": 60,
        "paddingRight": 60,
        "paddingTop": 30,
        "width": 750,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 35,
              "height": 70,
              "justifyContent": "center",
              "width": 70,
            },
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#555",
              "fontSize": 14,
            }
          }
        >
          Lap
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 1,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 35,
              "height": 70,
              "justifyContent": "center",
              "width": 70,
            },
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "fontSize": 14,
              },
              Object {
                "color": "#60B644",
              },
            ]
          }
        >
          Start
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "height": 1034,
        "paddingLeft": 15,
        "width": 750,
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
